R/xgb_binning.R

```diff
@@ -0,0 +1,250 @@
+#' Discretize numeric variables with XgBoost
+#'
+#' `step_xgb_binning` creates a *specification* of a recipe
+#'  step that will discretize numeric data (e.g. integeres or doubles)
+#'  into bins in a supervised way using an XgBoost model.
+#'
+#' @param recipe A recipe object. The step will be added to the sequence of
+#'   operations for this recipe.
+#' @param ... One or more selector functions to choose which variables are
+#'   affected by the step. See [selections()] for more details.
+#' @param role Defaults to `"predictor"`.
+#' @param trained A logical to indicate if the quantities for preprocessing
+#'   have been estimated.
+#' @param outcome A character string to indicate the outcome variable to train XbBoost models.
+#' @param rules The splitting rules of the best XgBoost tree to retain for each variable.
```

Comment:
It would be good to have the tuning parameters exposed (instead of hard-coded in `run_xgboost`) as main arguments to the step so make it easy to tune. This will add a lot of people will want to tune it.

Reply:
Then the following ones would be my set of candidates where we pass some reasonable defaults for this step:

eta (0.3)
max_bin (10, 20, 50) <- that's pretty much the most important parameter for this step
max_depth (1)
mind_child_weight (5)
colsample_bytree (0.3)

What do you think about the others such as nround and early_stopping rounds?

Reply:
I just saw your comment below on ```num_breaks``` - so would you expose more of them or only this main one that's critical to this step?

Reply:
I would err on the side of too many arguments. The tuning infrastructure coming up works best when the parameters that affect performance are main parameters. Also, Io would use `learn_rate` instead of `eta`; we try to avoid jargon.

Reply:
Point taken on ```learn_rate```. Also for the start, let's include 3 most critical parameters for this task which would be: ```max_bin```, ```eta``` and ```max_depth``` (with appropriately changed naming).